### ISSUE TITLE
Opentelemetry instrumentation for `graphql-batch`

### ISSUE THREAD

**Issue by toneymathews** [open-telemetry/opentelemetry-ruby-contrib#457]
TL;DR [GraphQL Batch](https://github.com/Shopify/graphql-batch) is a query batching executor for the [GraphQL Ruby](https://github.com/rmosolgo/graphql-ruby) gem. It’s used to solve n+1 query problems. As maintainers of GraphQL Batch, we would like to implement tracing so that a shared implementation may be available within [open-telemetry/opentelemtry-ruby-contrib](https://github.com/open-telemetry/opentelemetry-ruby-contrib).

## Defining the problem space and how the gem solves this problem
By default, every field in a GraphQL query is resolved individually on each object, in total isolation. Consider the following GraphQL query, assuming a SQL based data source with pagination omitted for simplicity:

```graphql
{
  orders(first: 50) {
    customer {
	name
    }
  }
}
```

On this query, first the `orders` field is resolved, which generates a SQL query for 50 orders from the database. Then for each individual order, the `customer` field is resolved, which generates an additional SQL query to load the customer for that order. In total, 51 SQL queries are executed (one to load the orders, and one per order to load the customer). This is an extremely costly way to resolve this data.

GraphQL Batch exposes a `GraphQL::Batch::Loader` [interface](https://github.com/Shopify/graphql-batch/blob/e476f7c25a975da4af39294f636c0c640eeb9dda/lib/graphql/batch/loader.rb). It exposes a `#perform` method that is used by a custom batch loader such as the CustomerLoader that aggregates the orders and performs a single SQL query.

```ruby
class CustomerLoader < GraphQL::Batch::Loader
  def perform(orders)
    customer_ids = orders.map(&:customer_id)
    # Query to fetch all customers.
    customers_by_id = Customers.where(id: customer_ids).index_by(&:id)

    orders.each do |order| 
      customer = customers_by_id[order.customer_id]
      fulfill(order, customer)  # For every order, fan out the corresponding customer.
    end
  end
end
```

Loaders based on `GraphQL::Batch::Loader` may be generic such as [RecordLoader](https://github.com/Shopify/graphql-batch/blob/e476f7c25a975da4af39294f636c0c640eeb9dda/examples/record_loader.rb) and [AssociationLoader](https://github.com/Shopify/graphql-batch/blob/master/examples/association_loader.rb) or a custom one such as the `CustomerLoader` above, for a specific purpose.

## What we propose
Add a batch loader within [open-telemetry/opentelemtry-ruby-contrib](https://github.com/open-telemetry/opentelemetry-ruby-contrib) that is traced so that a shared implementation is available. The perform method holds code that makes calls to remote data sources. We want to add tracing for the perform method to visualize and measure performance of a batch loader.

If it makes sense to do so, we can add a pull request to add instrumentation for `graphql-batch` as the next step for further feedback.

**Comment by arielvalentin** [comment]
Speaking for myself, adding instrumentation for `graphql-batch` would be a welcome addition to this repo.

I am looking forward to your contribution!

**Comment by arielvalentin** [comment]
@toneymathews Any chance you may have time to work on this?
